pkg/plugins/pg/destination/destination.go

```diff
@@ -16,114 +16,178 @@ package destination
+	if err != nil {
+		return fmt.Errorf("failed to get key: %w", err)
 	}
 
-	// keyCol holds the name of the column that records are keyed against.
-	var keyCol string
-	if r.Key != nil {
-		d, ok := r.Key.(record.StructuredData)
-		if !ok {
-			return s.Position, cerrors.Errorf("failed to parse key: %+v", r.Key)
-		}
+	keyColumnName := getKeyColumnName(key, d.config.keyColumnName)
 
-		// TODO: Handle composite keys
-		// Explicitly error out if we detect more than one key.
-		if len(d) > 1 {
-			return s.Position, cerrors.New("composite keys not yet supported")
-		}
+	tableName, err := d.getTableName(r.Metadata)
+	if err != nil {
+		return fmt.Errorf("failed to get table name for write: %w", err)
+	}
 
-		// add key data to the query
-		for key, val := range d {
-			keyCol = key
-			if _, ok := payload[key]; ok {
-				break
-			}
-			colArgs = append(colArgs, key)
-			valArgs = append(valArgs, val)
-		}
+	query, args, err := formatQuery(key, payload, keyColumnName, tableName)
+	if err != nil {
+		return fmt.Errorf("error formatting query: %w", err)
 	}
-	// check if the record has a table set. if it does, write the record to that
-	// table. if it doesn't, error out.
-	table, ok := r.Metadata["table"]
-	if !ok {
-		return s.Position, cerrors.New("record must provide a table name")
+
+	_, err = d.conn.Exec(ctx, query, args...)
+	if err != nil {
+		return fmt.Errorf("insert exec failed: %w", err)
 	}
 
-	// manually format the upsert and on conflict part of the query.
-	// the ON CONFLICT portion of this query needs to specify the constraint
-	// name. in our case, we can only rely on the record.Key's parsed
-	// StructuredData key.
-	// note: if other schema constraints prevent a write, this won't upsert on
-	// that conflict.
-	upsertQuery := fmt.Sprintf("ON CONFLICT (%s) DO UPDATE SET", keyCol)
+	return nil
+}
+
+func getPayload(r sdk.Record) (sdk.StructuredData, error) {
+	return structuredDataFormatter(r.Payload.Bytes())
+}
 
-	// range over the record's payload and create the upsert query and args
+func getKey(r sdk.Record) (sdk.StructuredData, error) {
+	return structuredDataFormatter(r.Key.Bytes())
+}
+
+func structuredDataFormatter(raw []byte) (sdk.StructuredData, error) {
+	if len(raw) == 0 {
+		return sdk.StructuredData{}, nil
+	}
+	data := make(map[string]interface{})
+	err := json.Unmarshal(raw, &data)
+	if err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
+// formatQuery manually formats the UPSERT and ON CONFLICT query statements.
+// the `ON CONFLICT` portion of this query needs to specify the constraint
+// name.
+// * In our case, we can only rely on the record.Key's parsed key value.
+// * If other schema constraints prevent a write, this won't upsert on
+// that conflict.
+func formatQuery(
+	key sdk.StructuredData,
+	payload sdk.StructuredData,
+	keyColumnName string,
+	tableName string,
+) (string, []interface{}, error) {
+	upsertQuery := fmt.Sprintf("ON CONFLICT (%s) DO UPDATE SET", keyColumnName)
```

Comment:
Sidenote: we need to rethink this statement, maybe only add it if there is a primary key, because it fails if there is no primary key on the target table. I'll create an issue for this.